src/embind/embind_ts.js

```diff
@@ -0,0 +1,276 @@
+      this.name = name;
+      this.returnType = returnType;
+      this.argumentTypes = argumentTypes;
+      this.thisType = thisType;
+    }
+
+    printSignature(nameMap, out) {
+      out.push('(');
+
+      let argOut = [];
+      for (let arg of this.argumentTypes) {
+        argOut.push(`${arg.name}: ${nameMap(arg.type)}`);
+      }
+      out.push(argOut.join(', '));
+      out.push(`): ${nameMap(this.returnType)}`);
+    }
+
+    printFunction(nameMap, out) {
+      out.push(`${this.name}`);
+      this.printSignature(nameMap, out);
+      out.push(';\n');
+    }
+
+    printModuleEntry(nameMap, out) {
+      out.push("  ");
+      this.printFunction(nameMap, out);
+    }
+  },
+  $ClassDefinition: class ClassDefinition {
+    constructor(typeId, name, base = null) {
+      this.typeId = typeId;
+      this.name = name;
+      this.methods = [];
+      this.constructors = [
+        new FunctionDefinition('default', this, [])
+      ];
+      this.base = base;
+    }
+
+    print(nameMap, out) {
+      out.push(`export interface ${this.name}`);
+      if (this.base) {
+        out.push(` extends ${this.base.name}`);
+      }
+      out.push(" {\n");
+      for (let method of this.methods) {
+        out.push("  ");
+        method.printFunction(nameMap, out);
+      }
+      out.push("  delete(): void;\n");
+      out.push("}\n\n");
+    }
+
+    printModuleEntry(nameMap, out) {
+      out.push(`  ${this.name}: {new`);
+      // TODO Handle constructor overloading
+      let constructor = this.constructors[this.constructors.length > 1 ? 1 : 0];
+      constructor.printSignature(nameMap, out);
+      out.push('};\n');
+    }
+  },
+  $TsPrinter: class TsPrinter {
+    constructor(definitions) {
+      this.definitions = definitions;
+      const jsString = 'ArrayBuffer|Uint8Array|Uint8ClampedArray|Int8Array|string';
+      this.builtInToJsName = new Map([
+        ["bool", "boolean"],
+        ["char", "number"],
+        ["unsigned char", "number"],
+        ["int", "number"],
+        ["unsigned int", "number"],
+        ["unsigned long", "number"],
+        ["float", "number"],
+        ["double", "number"],
+        ["void", "void"],
+        ["std::string", jsString],
+        ["std::basic_string<unsigned char>", jsString],
+        ["emscripten::val", "any"],
+      ]);
+    }
+
+    typeToJsName(type) {
+      if (type instanceof PrimitiveType) {
+        if (!this.builtInToJsName.has(type.name)) {
+          throw new Error(`Missing primitive type to TS type for "${type.name}"`);
+        }
+        return this.builtInToJsName.get(type.name)
+      }
+      return type.name;
+    }
+
+    print() {
+      var out = [];
+      for (let def of this.definitions) {
+        if (!def.print) {
+          continue;
+        }
+        def.print(this.typeToJsName.bind(this), out);
+      }
+      // Print module definitions
+      out.push("export interface MainModule {\n");
+      for (let def of this.definitions) {
```

Comment:
`const` here and for other loop variables in this file.   Actually `out` too..?

Reply:
I thought we were avoiding const/let? Has that changed?

Reply:
Thats is true in general, for output code that runs on the target.  But IIUC this can be considered compile time code?

And its really `... of ..` that that is more of a problem in generated code because our transpirer can't handle it yet.  

The code as written looks like its not output code since its contains the ` .. of .. ` so I assumed it was build time code.

Reply:
Yeah, I guess this is more compile time code since we'll only be running under node.